client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java

```diff
@@ -196,4 +229,19 @@ public static Position fromBytes(ByteBuffer buff) {
         return SERIALIZER.deserialize(new ByteArraySegment(buff));
     }
 
+    private synchronized void applySegmentOffsetUpdatesIfNeeded() {
```

Comment:
This lock is not guarding `updatesToSegmentOffsets` because it is shared by many objects.